### PR TITLE
Parser: handle newlines that are not in 'pre'

### PR DESCRIPTION
**Pull request by mightyiam**
Fix tests for this

### CODE REVIEW

File: src/test/unit/specific_feature_tests/lists.js [hunk 1]
```diff
@@ -2812,24 +2812,20 @@ test("Tab has no effect outside lists", function () {
...
+                    'Lorem ipsum dol\nor' +
                 '</li>' +
             '</ul>',
 
         expectedHtml = String() +
             '<ul>' +
                 '<li>' +
-                    'Lorem ipsum dolor sit amet, consectetur adipiscing elit.' +
-                    ' Integer sagittis porta dapibus.' +
+                    'Lorem ipsum dolor' +
```
Comment (mightyiam) [inline src/test/unit/specific_feature_tests/lists.js]: Not only in lists, but everywhere that is not a `pre`, actually.

**Comment by mightyiam** [comment]
What I'm concerned about in this approach is whether some paste from word things work as expected or not. Perhaps Word provides some linebreaks which actually should be interpreted as `br` tags?

**Comment by mightyiam** [comment]
If we choose to proceed with this change, then I would write two more tests:

- [ ] newlines are stripped in a few different elements
- [ ] newlines are kept in `pre` elements

File: src/wymeditor/parser/xhtml-sax-listener.js [hunk 2]
```diff
@@ -284,13 +284,10 @@ WYMeditor.XhtmlSaxListener.prototype.getTagForStyle = function (style) {
...
-        // We should strip trailing newlines from text inside li tags because
-        // IE adds random significant newlines inside nested lists
-        text = text.replace(/(\r|\n|\r\n)+$/g, '');
-
-        // Let's also normalize multiple newlines down to a single space
-        text = text.replace(/(\r|\n|\r\n)+/g, ' ');
+    if (!(this.last_tag && this.last_tag === 'pre')) {
+        // We respect newlines only when in 'pre'. Outside of 'pre',
+        // we remove all of them.
+        text = text.replace(/(\r|\n|\r\n)/g, '');
```
Comment (winhamwr) [inline src/wymeditor/parser/xhtml-sax-listener.js]: We can't just remove them. A single newline is significant.

These two things are actually different:

`<p>hello\nshahar</p>`

`<p>helloshahar</p>`

**Comment by mightyiam** [comment]
The last commit introduces a solution that seems excellent to me. It also fixes the failing tests from #685.

From the commit:
``` js
if (!(this.last_tag && this.last_tag === 'pre')) {
    // We respect newlines as they are, only when in 'pre'
    // outside of 'pre',
    // at end of elements they are removed
    text = text.replace(/(\r|\n|\r\n)+$/g, '');
    // the rest are replaced with spaces
    text = text.replace(/(\r|\n|\r\n)/g, ' ');
}
```

And here is what it looks like:

## Chrome
screenshot from 2015-03-06 20 38 21

## IE8
screenshot from 2015-03-06 20 39 26

If you OK this I'll write those two tests.

**Comment by mightyiam** [comment]
Please see this behavior as an improvement over https://github.com/wymeditor/wymeditor/issues/685#issuecomment-77609999.

**Comment by winhamwr** [comment]
I knew I felt weird about this and I finally figured out why. The way that whitespace is handled within an element is actually defined in the [html5 spec](http://www.w3.org/TR/CSS2/text.html#white-space-model). The spec makes it clear that the rendered behavior of whitespace can be controlled with the [white-space](https://developer.mozilla.org/en-US/docs/Web/CSS/white-space) css property.

That means that, at the HTML-parsing level, we can't safely remove whitespace arbitrarily. For example, it wouldn't be rare to use a `code {white-space: pre}` CSS declaration to make rendering code samples whitespace-sensitive.

**Pull request closed without merging**